rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,621 @@
+
+#### By clicking “Remove agent” on the status tab
+
+This removes the config and the data dir together with any CMC-related app state and an attempt is
+made to remove the CMC role from the cluster. The agent binary is kept in the sessionData folder to
+enable a scenario where the user wants to remove the agent from one cluster but wants to keep using
+it in another cluster.
+
+Removing the agent performs the following actions:
+
+1. Stop the agent process for the given cluster.
+2. Capture the UUID of the node for use in subsequent steps.
+3. Remove the agent data at `<userData>/agents/<profile>`.
+4. Call [`DeleteNode`](https://github.com/gravitational/teleport/blob/7e174f9aaf17d0f3ea08740aac42da49eb7ec8da/api/client/client.go#L2044)
+   with the node UUID to remove the node from the cluster.
+
+    The request may fail if the user does not have permission to remove nodes. This will not be
+    considered a fatal error. Instead, at the end of the agent removal process Connect will explain
+    to the user that the node can still be visible for a few minutes until it is purged from the
+    cluster cache. This should help prevent user confusion regarding the continued visibility of the
+    CMC node, as they may not be fully aware of how the cluster cache works.
+
+5. Remove the CMC role.
+
+    The CMC role should be removed from the current user and then an attempt should be made to
+    remove the role itself. This can fail if the role has since been set for other users. If the
+    attempt fails, we should log the error and continue.
+
+6. Remove CMC-related app state.
+
+    If the user made a connection to the CMC node, in the top left in the connections list they are
+    going to see a connection for that node. Since the node will no longer be accessible, it doesn’t
+    make sense to keep it around.
+
+    Connect also has to remove the `autoStart` flag.
+
+
+#### By logging out
+
+On logout, we have to remove the config file and the data dir. Imagine a situation where the user
+logs in as the cluster user foo, sets up CMC, then logs out and logs in as bar. If we kept the
+config file and the data dir on logout, the user would now be logged in as bar with CMC configured
+for foo.
+
+To avoid conflicts and keep things simple, the config file and the data dir will be treated as any
+other piece of local state that is bound to a specific profile and thus will be removed on logout.
+
+Any workspace-specific app state related to CMC (such as the `autoStart` flag) will be wiped out on
+logout as well.
+
+The downside of not doing any extra cleanup on logout is that if the user runs CMC, then logs out
+and immediately logs back in, the cache will still contain the now gone node. This is a rough edge
+we’re willing to accept in the MVP over adding more complexity to handle this specific scenario.
+
+### Availability of the feature
+
+The feature is supported only on macOS and Linux because the Teleport SSH agent does not work on
+Windows.
```

Comment:
@zmb3 @gzdunek Thoughts on supporting Connect My Computer on Windows through WSL?

I don't know why we haven't considered WSL all this time, I think it's a viable option.

I just ran a node agent through WSL and it appears to be working just fine. There's a few failing unit tests, but they might as well be due to me not interacting with Windows firewall prompts fast enough.

The FAQ doesn't list [any downside between WSL and a regular Linux distro](https://learn.microsoft.com/en-us/windows/wsl/faq#can-i-use-wsl-for-production-scenarios--) which would be a showstopper for us. [The WSL team did some tests](https://learn.microsoft.com/en-gb/archive/blogs/wsl/testing-the-windows-subsystem-for-linux) to ensure WSL compatibility.

The only downside is that setting up WSL is an interactive process done through CLI. You need to select a distro (which we can preselect for you) and provide a password for the new unix account. But we can just open a new tab with a terminal, let the user set up WSL and then continue the regular Connect My Computer setup. I bet some part of the target audience will even have WSL installed.

I think we could spend a little bit more time on the MVP if it means reaching twice as much users. We could first build a version that works on Unix only and then see how we can fit it into WSL.

Reply:
If it works, I'm all for supporting it. 

>The only downside is that setting up WSL is an interactive process done through CLI. You need to select a distro (which we can preselect for you) and provide a password for the new unix account.

Is this accurate? AFAIK [the distro is preselected be default](https://learn.microsoft.com/en-us/windows/wsl/setup/environment) (Ubuntu) and then you have to provide a username and password for the account (but I failed to verify that on a Parallels VM, it looks like WSL2 doesn't work there).

>I think we could spend a little bit more time on the MVP if it means reaching twice as much users. We could first build a version that works on Unix only and then see how we can fit it into WSL.

+1
It seems to me that after setting up WSL, we would have to take care of two things:
- spawning `wsl teleport` instead of `teleport`,
- making the mechanism that checks if the parent PID hasn't changed work on Windows.

But probably there would be something more :)

Reply:
> Is this accurate? AFAIK [the distro is preselected be default](https://learn.microsoft.com/en-us/windows/wsl/setup/environment) (Ubuntu) and then you have to provide a username and password for the account

Oh, I've been installing WSL on a couple of machines during Rails Girls and the guides explicitly specified Ubuntu. But you still need to provide a password, so the install is still interactive.

There seems to be a way to set up WSL without user interaction (https://github.com/microsoft/WSL/issues/3369#issuecomment-803522089). But as WSL is a system-wide thing and not specific to Teleport or Connect, I'd rather have the user install it in a separate interactive step. Running the interactive installer would also show the progress bars and errors OOTB.

> but I failed to verify that on a Parallels VM, it looks like WSL2 doesn't work there

Yeah, you can downgrade it to WSL1 for now and use it with Parallels, that's how I tested everything.

> * making the mechanism that checks if the parent PID hasn't changed work on Windows

There's a couple of issues mentioning some inconsistencies between WSL1 and WSL2, but there's a slight chance that on Windows we'll be able to use general Unix behavior when it comes to PID monitoring, without having to implement Windows-specific handling.

* https://github.com/microsoft/WSL/issues/2854
* https://github.com/winsiderss/systeminformer/issues/153#issuecomment-462079170